river/src/main/java/org/jboss/marshalling/river/SerializableClassDescriptor.java

```diff
@@ -54,10 +56,10 @@ public String toString() {
 
     public Class<?> getNonSerializableSuperclass() {
         final ClassDescriptor descriptor = getSuperClassDescriptor();
-        if (descriptor instanceof SerializableClassDescriptor) {
+        final Class<?> type = descriptor.getType();
+        if (type == null) return Object.class;
```

Comment:
A null type only means the descriptor corresponds to a non-present class, it doesn't mean that the supertype is `Object`.    There could be a serializable superclass that is yet present, whose superclass is in turn not `Object` and also not serializable.

Reply:
Apparently that scenario is not tested and the current implementation forms a regression.
This is the easiest means to fix that regression without introducing side-effects (that would be caught by existing tests).

Reply:
Maybe something like this is safer/easier: https://github.com/dmlloyd/jboss-marshalling/commit/example

Reply:
That doesn't pass my test case.

Reply:
Hmm I see, then we have two - or maybe three - problems.

If the class is marked as serializable on the wire but it's not locally serializable, we need to discard its serialized information - something we don't have a facility for at the moment.  We do support the opposite - so called "gap" classes - where the local side has extra classes not present on the sending side, but that doesn't help us here I don't think.

We might be able to treat it as non-present in this case though, when reading ID_SERIALIZABLE_CLASS/ID_WRITE_OBJECT_CLASS, by checking the resolved class to see if it is serializable, and if not, use "null" to indicate non-present.  Or by simply rejecting the stream; I'm not sure the spec requires this behavior (will check).

Reply:
https://docs.oracle.com/javase/8/docs/platform/serialization/spec/version.html#a5172 says:

> Removing either Serializable or Externalizable is an incompatible change since when written it will no longer supply the fields needed by older versions of the class.

So we could perhaps reject this case.

Reply:
Although later at https://docs.oracle.com/javase/8/docs/platform/serialization/spec/version.html#a6754 we have:

> Removing classes - Comparing the class hierarchy in the stream with that of the current class can detect that a class has been deleted. In this case, the fields and objects corresponding to that class are read from the stream. Primitive fields are discarded, but the objects referenced by the deleted class are created, since they may be referred to later in the stream. They will be garbage-collected when the stream is garbage-collected or reset.

So an argument could be made that we should discard as if the class were missing.

Reply:
Here's a new version (just inline for simplicity):

```patch
diff --git a/river/src/main/java/org/jboss/marshalling/river/RiverUnmarshaller.java b/river/src/main/java/org/jboss/marshalling/river/RiverUnmarshaller.java
index 3303ce5..a30832f 100644
--- a/river/src/main/java/org/jboss/marshalling/river/RiverUnmarshaller.java
+++ b/river/src/main/java/org/jboss/marshalling/river/RiverUnmarshaller.java
@@ -1023,6 +1023,9 @@ public class RiverUnmarshaller extends AbstractUnmarshaller {
                 } catch (ClassNotFoundException cnfe) {
                     if (required) throw cnfe;
                 }
+                if (! serializabilityChecker.isSerializable(clazz) && required) {
+                    throw new ClassNotFoundException(className);
+                }
                 final FutureSerializableClassDescriptor descriptor = new FutureSerializableClassDescriptor(clazz, classType);
                 classCache.set(idx, descriptor);
                 final int cnt = readInt();
diff --git a/river/src/main/java/org/jboss/marshalling/river/SerializableClassDescriptor.java b/river/src/main/java/org/jboss/marshalling/river/SerializableClassDescriptor.java
index ae82b9f..dba9b2e 100644
--- a/river/src/main/java/org/jboss/marshalling/river/SerializableClassDescriptor.java
+++ b/river/src/main/java/org/jboss/marshalling/river/SerializableClassDescriptor.java
@@ -57,7 +57,7 @@ abstract class SerializableClassDescriptor extends ClassDescriptor {
         if (descriptor instanceof SerializableClassDescriptor) {
             return ((SerializableClassDescriptor) descriptor).getNonSerializableSuperclass();
         } else {
-            return descriptor.getType();
+            return descriptor.getNearestType();
         }
     }
 }
```

Reply:
The theory here is that if the given class is not locally serializable, we treat it as a "skip".  This isn't exactly perfect though since the local class' constructor still won't get called.

Reply:
This version might be it.  In this case we ensure that the local nonserializable superclass is considered to be the superclass to instantiate even if a class with the same name is on the wire.  In theory.

```patch
diff --git a/river/src/main/java/org/jboss/marshalling/river/RiverUnmarshaller.java b/river/src/main/java/org/jboss/marshalling/river/RiverUnmarshaller.java
index 3303ce5..3303cea 100644
--- a/river/src/main/java/org/jboss/marshalling/river/RiverUnmarshaller.java
+++ b/river/src/main/java/org/jboss/marshalling/river/RiverUnmarshaller.java
@@ -1023,7 +1023,11 @@ public class RiverUnmarshaller extends AbstractUnmarshaller {
                 } catch (ClassNotFoundException cnfe) {
                     if (required) throw cnfe;
                 }
-                final FutureSerializableClassDescriptor descriptor = new FutureSerializableClassDescriptor(clazz, classType);
+                final boolean localSerializable = clazz != null && serializabilityChecker.isSerializable(clazz);
+                if (! localSerializable && required) {
+                    throw new ClassNotFoundException(className);
+                }
+                final FutureSerializableClassDescriptor descriptor = new FutureSerializableClassDescriptor(localSerializable ? clazz : null, classType);
                 classCache.set(idx, descriptor);
                 final int cnt = readInt();
                 final String[] names = new String[cnt];
@@ -1036,7 +1040,7 @@ public class RiverUnmarshaller extends AbstractUnmarshaller {
                 }
                 ClassDescriptor superDescriptor = doReadClassDescriptor(readUnsignedByte(), false);
                 final Class<?> superClazz = clazz == null ? superDescriptor.getNearestType() : clazz.getSuperclass();
-                if (superDescriptor != null) {
+                if (superDescriptor != null && (clazz == null || localSerializable)) {
                     final Class<?> superType = superDescriptor.getNearestType();
                     if (clazz != null && ! superType.isAssignableFrom(clazz)) {
                         throw new InvalidClassException(clazz.getName(), "Class does not extend stream superclass");
diff --git a/river/src/main/java/org/jboss/marshalling/river/SerializableClassDescriptor.java b/river/src/main/java/org/jboss/marshalling/river/SerializableClassDescriptor.java
index ae82b9f..dba9b2e 100644
--- a/river/src/main/java/org/jboss/marshalling/river/SerializableClassDescriptor.java
+++ b/river/src/main/java/org/jboss/marshalling/river/SerializableClassDescriptor.java
@@ -57,7 +57,7 @@ abstract class SerializableClassDescriptor extends ClassDescriptor {
         if (descriptor instanceof SerializableClassDescriptor) {
             return ((SerializableClassDescriptor) descriptor).getNonSerializableSuperclass();
         } else {
-            return descriptor.getType();
+            return descriptor.getNearestType();
         }
     }
 }
```

Reply:
Doesn't pass my test case.
This is all ghost hunting, which is likely repeated in the future as there is no test coverage for these scenarios.
Additional tests are needed (which may pop new bugs / regressions), which I consider something for another day.